parity-crypto/src/secp256k1_alt.rs

```diff
@@ -0,0 +1,411 @@
+	}
+
+}
+
+impl SecretKeyTrait for SecretKey {
+	type VecRepr = ClearOnDrop<Vec<u8>>;
+
+	fn to_vec(&self) -> Self::VecRepr {
+		ClearOnDrop::new(self.0.serialize().to_vec())
+	}
+
+	fn sign(&self, message: &[u8]) -> Result<Vec<u8>, Error> {
+
+	 	let mut buf = [0;32];
+		buf.copy_from_slice(&message[..]); // panic on incorrect message size
+		let message = Message::parse(&buf);
+		let (sig, rec_id) = secp256k1::sign(&message, &self.0)?;
+		let mut data_arr = vec![0; 65];
+		data_arr[0..64].copy_from_slice(&sig.serialize());
+		data_arr[64] = rec_id.serialize();
+		Ok(data_arr)
+	}
+
+}
+
+/// random secret key for rand 0.5
+fn random_sec<R: Rng>(rng: &mut R) -> SecretKey {
+	loop {
+		let mut ret = [0u8; 32];
+		rng.fill_bytes(&mut ret);
+
+		match Secp256k1::secret_from_slice(&ret) {
+			Ok(key) => return key,
+			Err(_) => (),
+		}
+	}
+}
+
+impl FixAsymSharedSecret for SecretKey {
+	type Other = PublicKey;
+	type Result = SharedSecretAsRef;
+
+	fn shared_secret(&self, publ: &Self::Other) -> Result<Self::Result, Error> {
+		let shared = secp256k1::SharedSecret::new(&publ.0, &self.0)?;
+		Ok(SharedSecretAsRef(shared))
+	}
+
+}
+
+#[derive(PartialEq, Eq, Debug, Clone)]
+pub struct SharedSecretAsRef(pub secp256k1::SharedSecret);
+
+impl AsRef<[u8]> for SharedSecretAsRef {
+	fn as_ref(&self) -> &[u8] {
+		self.0.as_ref()
+	}
+}
+
+fn aff_to_public(aff_pub: &mut Affine) -> Result<PublicKeyInner, Error> {
+	let mut buff = [4;65];
+	let mut buff2 = [0;32];
+	aff_pub.x.normalize();
+	aff_pub.x.fill_b32(&mut buff2);
+	buff[1..33].copy_from_slice(&buff2[..]);
+	aff_pub.y.normalize();
+	aff_pub.y.fill_b32(&mut buff2);
+	buff[33..65].copy_from_slice(&buff2[..]);
+	Ok(PublicKeyInner::parse(&buff)?)
+}
+
+struct SecretScalar(pub Scalar);
+
+impl Drop for SecretScalar {
+	fn drop(&mut self) {
+		self.0.clear();
+	}
+}
+
+/* private inner method but this would avoid a scalar instantiation
+fn mul_in_place_scalar(a: &mut Scalar, b: &Scalar) {
+	let mut l = [0u32; 16];
+	a.mul_512(b, &mut l);
+	a.reduce_512(&l);
+}
+*/
+
+impl FiniteField for Secp256k1 {
+
+	fn generator_x() -> &'static[u8] { &GENERATOR_X[..] }
+	fn generator_y() -> &'static[u8] { &GENERATOR_Y[..] }
+	fn curve_order() -> &'static[u8] { &CURVE_ORDER[..] }
+
+	fn public_mul(pub_key: &mut Self::PublicKey, sec_key: &Self::SecretKey) -> Result<(), Error> {
+		let sec_scal = SecretScalar(sec_key.0.clone().into());
+		let mut pub_aff: Affine = pub_key.0.clone().into();
+		let mut pub_jac = Jacobian::default();
+		pub_jac.set_ge(&pub_aff);
+		//ECMULT_GEN_CONTEXT.ecmult_gen(&mut pub_jac, &sec_scal);
+		//pub_aff.set_gej(&pub_jac);
+		let mut zero = Scalar::default();
+		zero.set_int(0);
+		let mut res = Jacobian::default();
+		ECMULT_CONTEXT.ecmult(&mut res, &pub_jac, &sec_scal.0, &zero);
+		pub_aff.set_gej(&res);
+		*pub_key = PublicKey::new(aff_to_public(&mut pub_aff)?);
+		Ok(())
+	}
+
+	fn public_add(pub_key: &mut Self::PublicKey, other_public: &Self::PublicKey) -> Result<(), Error> {
```

Comment:
Add this in `libsecp256k1` as `PublicKey::combine`.

Reply:
My tests are failing with combine, I think it's related to an error when manipulating infinite point for first addition.

Reply:
It's a bug in `add_ge`. Sorry! Fixed and published `0.2.1`.